File: obdiggio-android/core/src/main/kotlin/org/diggio/obdiggio/core/obd/Transport.kt
package org.diggio.obdiggio.core.obd

/**
 * Canale byte bidirezionale tra il livello ELM327 e il mezzo fisico.
 *
 * Le implementazioni (BLE su Android, oppure il simulatore) chiamano [feed]
 * quando ricevono byte; il livello ELM327 legge con [readUntil] fino al prompt
 * `>`. Il buffer di ricezione è thread-safe perché su BLE le notifiche arrivano
 * su un thread diverso da quello che invia i comandi.
 */
abstract class Transport {

    private val buffer = ArrayDeque<Byte>()
    private val lock = Any()

    abstract fun open()
    abstract fun close()
    abstract fun write(data: ByteArray)
    abstract val isConnected: Boolean

    /** Le implementazioni chiamano questo metodo quando arrivano byte. */
    protected fun feed(data: ByteArray) {
        synchronized(lock) { data.forEach { buffer.addLast(it) } }
    }

    /** Svuota il buffer di ricezione. */
    fun clear() {
        synchronized(lock) { buffer.clear() }
    }

    /**
     * Legge dal buffer finché non trova [terminator] o scade [timeoutMs].
     * Restituisce i byte accumulati (terminatore incluso se trovato); in timeout
     * restituisce comunque quanto ricevuto.
     */
    fun readUntil(terminator: Byte = '>'.code.toByte(), timeoutMs: Long = 5000): ByteArray {
        val deadline = System.nanoTime() + timeoutMs * 1_000_000
        while (System.nanoTime() < deadline) {
            synchronized(lock) {
                val idx = buffer.indexOf(terminator)
                if (idx != -1) {
                    val out = ByteArray(idx + 1) { buffer.removeFirst() }
                    return out
                }
            }
            Thread.sleep(10)
        }
        synchronized(lock) {
            val out = buffer.toByteArray()
            buffer.clear()
            return out
        }
    }
}

File: obdiggio/obdiggio/obd/transport.py
"""Astrazione di trasporto tra il livello ELM327 e il canale fisico.

Un ELM327 su BLE espone (tipicamente) una caratteristica di *notify* da cui
arrivano i byte e una caratteristica di *write* su cui inviare i comandi. I dati
in ricezione arrivano quindi in modo asincrono, spezzettati in piu' notifiche.

:class:`Transport` normalizza tutto questo: le sottoclassi chiamano
``_feed(data)`` quando ricevono byte, e il livello ELM327 usa
``read_until()`` per leggere in modo bloccante fino a un terminatore
(il prompt ``>`` dell'ELM327).
"""

from __future__ import annotations

import threading
import time
from abc import ABC, abstractmethod


class Transport(ABC):
    """Canale byte bidirezionale con buffer di ricezione thread-safe."""

    def __init__(self) -> None:
        self._buffer = bytearray()
        self._lock = threading.Lock()

    # --- API che le sottoclassi devono implementare -----------------------

    @abstractmethod
    def open(self) -> None:
        """Apre/attiva il canale (connessione BLE, ecc.)."""

    @abstractmethod
    def close(self) -> None:
        """Chiude il canale."""

    @abstractmethod
    def write(self, data: bytes) -> None:
        """Invia byte sul canale fisico."""

    @property
    @abstractmethod
    def is_connected(self) -> bool:
        ...

    # --- API usata dal livello ELM327 -------------------------------------

    def _feed(self, data: bytes) -> None:
        """Le sottoclassi chiamano questo metodo quando arrivano byte."""
        with self._lock:
            self._buffer.extend(data)

    def clear(self) -> None:
        """Svuota il buffer di ricezione."""
        with self._lock:
            self._buffer.clear()

    def read_until(self, terminator: bytes = b">", timeout: float = 5.0) -> bytes:
        """Legge dal buffer finche' non trova ``terminator`` o scade il timeout.

        Ritorna i byte accumulati (terminatore incluso se trovato). In timeout
        ritorna comunque quanto ricevuto finora.
        """
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            with self._lock:
                idx = self._buffer.find(terminator)
                if idx != -1:
                    end = idx + len(terminator)
                    out = bytes(self._buffer[:end])
                    del self._buffer[:end]
                    return out
            time.sleep(0.01)
        # Timeout: restituisci quanto abbiamo.
        with self._lock:
            out = bytes(self._buffer)
            self._buffer.clear()
        return out
